storage/src/index/storage.rs

```diff
@@ -18,63 +24,202 @@ pub struct ValueIterator<'a, V> {
+    /// Insert a new value at the front of the list. We always add to the front so behavior is
+    /// consistently last in, first out. This means most recently added values will be returned
+    /// first by the iterators, providing an LRU like behavior.
+    pub fn insert(&mut self, mut value: V) {
+        let entry = self.entry.take().unwrap();
+
+        let mut occupied_entry = match entry {
+            OccupiedOrVacant::Occupied(occupied_entry) => occupied_entry,
+            OccupiedOrVacant::Vacant(vacant_entry) => {
+                // Key had no associated values, so just turn the vacant entry into an occupied one.
+                let record = Record { value, next: None };
+                let occupied_entry = vacant_entry.insert_entry(record);
+                self.entry = Some(OccupiedOrVacant::Occupied(occupied_entry));
+                return;
+            }
+        };
+
+        // This key already has a value, so add the new value to the front of the list.
+        let record = occupied_entry.get_mut();
+        swap(&mut record.value, &mut value); // puts the new value at the front
+        record.next = Some(Box::new(Record {
+            value,
+            next: record.next.take(),
+        }));
+        self.entry = Some(OccupiedOrVacant::Occupied(occupied_entry));
+
+        self.collisions_counter.inc();
+    }
+}
+
+/// An iterator over all values associated with a translated key, allowing for mutation and deletion
+/// of the current element.
+pub struct DeleteValueIterator<'a, K, V> {
+    prev: Option<*mut Record<V>>,
+    next: Option<*mut Record<V>>,
+    entry: Option<OccupiedEntry<'a, K, Record<V>>>,
+    pruned_counter: &'a Counter,
+}
+
+/// DeleteValueIterator must be sendable across threads so it can be held across a journal's read
+/// async boundary.
+unsafe impl<K, V> Send for DeleteValueIterator<'_, K, V> {}
+
+impl<'a, K, V> Iterator for DeleteValueIterator<'a, K, V> {
     type Item = &'a mut V;
 
     fn next(&mut self) -> Option<Self::Item> {
-        match self.next.take() {
+        match self.next {
             Some(next) => {
-                let loc = &mut next.value;
-                self.next = next.next.as_deref_mut();
+                let current = unsafe { &mut (*next) };
+                let loc = &mut current.value;
+                self.prev = self.next;
+                self.next = current
+                    .next
+                    .as_mut()
+                    .map(|next_next| next_next.as_mut() as *mut _);
                 Some(loc)
             }
             None => None,
         }
     }
 }
 
+impl<K, V> DeleteValueIterator<'_, K, V> {
+    /// Remove the value last returned from this iterator from the map. If no value has been
+    /// returned yet, or if the last returned value was deleted already, then this is a no-op.
+    pub fn remove(&mut self) {
+        // This implementation is linear in the length of the linked list since it searches the list
```

Comment:
To implement this, we could just do something like `self.prev_prev` right (where we only set it if self.prev is not `None`)?

Reply:
yep.

Reply:
went ahead and added this as discussed IRL